File: Nostos.Shared/Dtos/ProviderDtos.cs
using System;
using System.Collections.Generic;

namespace Nostos.Shared.Dtos;

// --- EXTERNAL CONTENT PROVIDERS (issue #166) ---
// The wire shape of the provider/acquisition boundary. Deliberately its own
// small vocabulary rather than a reuse of the library DTOs: these describe
// something that is NOT yet a Nostos book, and keeping them separate is what
// stops provider concepts from spreading into the library contract.
//
// No provider-specific field appears here (no GutenbergId, no LibriVoxId) — a
// provider normalizes its catalogue into exactly these shapes.

/// <summary>One selectable source in the "add from a source" surface.</summary>
public sealed record ProviderSummaryDto(
    string Id,
    string DisplayName,
    IReadOnlyList<string> Capabilities,
    /// <summary>The source's own rights wording, shown as-is. Never a Nostos claim.</summary>
    string? RightsNotice);

public sealed record ProviderAssetDto(
    string Id,
    /// <summary>"ebook" or "audiobook".</summary>
    string Kind,
    string Label,
    string? SourceFormat,
    long? SizeBytes,
    bool IsPreferred);

public sealed record ProviderItemDto(
    string ProviderId,
    string ExternalId,
    /// <summary>"ebook" or "audiobook", available on thin search results.</summary>
    string MediaKind,
    string Title,
    string? Subtitle,
    string? Author,
    string? Description,
    string? Language,
    string? Publisher,
    string? PublishedDate,
    string? Categories,
    string? Narrator,
    string? Duration,
    int? PageCount,
    IReadOnlyList<ProviderAssetDto> Assets,
    /// <summary>Nostos-relative URL that proxies the artwork, or null when the source has none.</summary>
    string? CoverUrl,
    string? SourceUrl,
    string? RightsStatement,
    /// <summary>Source parts/tracks, e.g. the number of LibriVox sections.</summary>
    int? PartCount);

public sealed record ProviderSearchResultDto(
    IReadOnlyList<ProviderItemDto> Items,
    bool HasMore,
    /// <summary>Provider-supplied note about a thin result set (e.g. a catalogue that only matches whole titles).</summary>
    string? Notice);

/// <summary>Per-provider outcome of one aggregate discovery request.</summary>
public sealed record ProviderDiscoverySourceStatusDto(
    string ProviderId,
    string DisplayName,
    bool Succeeded,
    string? Notice,
    string? ErrorCode);

/// <summary>
/// Provider-neutral discovery response. Successful results remain useful when
/// one source is temporarily unavailable.
/// </summary>
public sealed record ProviderDiscoverySearchResultDto(
    IReadOnlyList<ProviderItemDto> Items,
    bool HasMore,
    IReadOnlyList<ProviderDiscoverySourceStatusDto> Sources);

/// <summary>
/// The user's own edits to the metadata an import would otherwise take from the
/// source verbatim, for an import started from a prefilled form.
///
/// Every field means the same thing: `null` (or absent) keeps what the source
/// says, and a value replaces it. An empty string clears the field — the user
/// emptied that box, and putting the source's text back would be worse than
/// dropping it. `Title` is the exception: it falls back to the source, because a
/// book with no title is not a book.
///
/// Send only what the user actually changed. A field left null still follows the
/// source, so a provider-side correction is never overwritten by a stale value
/// the client echoed back.
/// </summary>
public sealed record ProviderMetadataOverridesDto(
    string? Title = null,
    string? Subtitle = null,
    string? Author = null,
    string? Description = null,
    string? Language = null,
    string? Publisher = null,
    string? PublishedDate = null,
    string? Categories = null,
    string? Narrator = null,
    string? Duration = null,
    int? PageCount = null);

/// <summary>
/// Start an import. Note there is no URL field: a client names a provider, an
/// item and optionally an asset, and the server resolves the rest.
/// </summary>
public sealed record ProviderAcquireRequestDto(
    string ExternalId,
    string? AssetId = null,
    IReadOnlyList<Guid>? CollectionIds = null,
    bool IncludeCover = true,
    /// <summary>
    /// Optional. Present only when the client showed the user the metadata first
    /// and they were allowed to change it.
    /// </summary>
    ProviderMetadataOverridesDto? MetadataOverrides = null);

public sealed record ProviderAcquisitionDto(
    string JobId,
    /// <summary>queued | running | succeeded | failed | cancelled</summary>
    string State,
    string Stage,
    int Percent,
    string? Detail,
    string ProviderId,
    string ExternalId,
    string? AssetId,
    Guid? BookId,
    string? ErrorCode,
    string? Message,
    DateTime CreatedAt,
    DateTime UpdatedAt);

File: Nostos.Frontend/src/app/core/dtos/provider.dtos.ts
// --- EXTERNAL CONTENT PROVIDERS (issues #166 / #167) ---
// The wire shape of the provider/acquisition boundary. Deliberately its own
// small vocabulary rather than a reuse of the library DTOs: these describe
// something that is NOT yet a Nostos book, which is what keeps provider
// concepts out of the library contract.
//
// Note there is no URL field anywhere a client can *send*: the client names a
// provider, an item and an asset, and the server resolves the location.

export interface ProviderSummary {
  id: string;
  displayName: string;
  /** Lower-case capability names, e.g. ['search', 'itemretrieval', 'ebookacquisition']. */
  capabilities: string[];
  /** The source's own rights wording, shown as-is. Never a Nostos claim. */
  rightsNotice: string | null;
}

export interface ProviderAsset {
  id: string;
  kind: 'ebook' | 'audiobook' | string;
  label: string;
  sourceFormat: string | null;
  sizeBytes: number | null;
  isPreferred: boolean;
}

export interface ProviderItem {
  providerId: string;
  externalId: string;
  title: string;
  subtitle: string | null;
  author: string | null;
  description: string | null;
  language: string | null;
  publisher: string | null;
  publishedDate: string | null;
  categories: string | null;
  narrator: string | null;
  duration: string | null;
  pageCount: number | null;
  assets: ProviderAsset[];
  /** Nostos-relative URL that proxies the artwork, or null when the source has none. */
  coverUrl: string | null;
  sourceUrl: string | null;
  rightsStatement: string | null;
  /** Source parts/tracks, e.g. the number of sections of a LibriVox recording. */
  partCount: number | null;
}

export interface ProviderSearchResult {
  items: ProviderItem[];
  hasMore: boolean;
  /**
   * A provider-supplied note about a thin result set. Several catalogues can
   * only match a whole title or an author surname, so an unexplained empty list
   * would read as a broken search.
   */
  notice: string | null;
}

export interface ProviderAcquireRequest {
  externalId: string;
  assetId?: string | null;
  collectionIds?: string[] | null;
  includeCover?: boolean;
}

export type AcquisitionState = 'queued' | 'running' | 'succeeded' | 'failed' | 'cancelled';

export interface ProviderAcquisition {
  jobId: string;
  state: AcquisitionState;
  /** queued | starting | resolving | checking | downloading | validating | assembling | importing | done | failed */
  stage: string;
  percent: number;
  detail: string | null;
  providerId: string;
  externalId: string;
  assetId: string | null;
  bookId: string | null;
  errorCode: string | null;
  message: string | null;
  createdAt: string;
  updatedAt: string;
}

export const ACQUISITION_FINISHED_STATES: ReadonlySet<AcquisitionState> = new Set([
  'succeeded',
  'failed',
  'cancelled',
]);
